home/gettingstarted/0.7/javausage.md

```diff
@@ -12,66 +12,117 @@ outof: 50
-            .put("exec.benchRuns", 20)
-            .put("exec.independentSamples", 1)
-            .put("exec.outliers.covMultiplier", 1.5)
-            .put("exec.outliers.suspectPercent", 40);
-
-Next, we declare a generator for the `groupBy` class (see more about generators in the [generator section](/home/gettingstarted/0.7/generators/index.html)):
-
-          public JavaGenerator<LinkedList<Integer>> generator() {
-            JavaGenerator<Integer> sizes = new SingleGen("size", 5000000);
-            return new CollectionGenerators(sizes).lists();
+In the Scala frontend, we encode the groups using nested closures.
+In the Java frontend, we encode them by placing the `benchmark` annotation 
+that argument denotes the group.
+We now declare benchmark snippet stub in a `list` group:
+
+      @benchmark("list")
+      public void histogram() {
+      
+      }
+      
+Next, we want a custom context:
+
+      public final Context config = new ContextBuilder()
+          .put("exec.benchRuns", 20)
+          .put("exec.independentSamples", 1)
+          .put("exec.outliers.covMultiplier", 1.5)
+          .put("exec.outliers.suspectPercent", 40)
+          .build();
+          
+We want it not only for our histogram method 
+but for all benchmark snippets in the `list` group:
+    
+    @scopes({
+      @scopeCtx(scope = "list", context = "config")
+    })
+    public class MyHistogramTest extends JBench.OfflineReport {
+
+Next, we declare a generator for our snippet (see more about generators in the [generator section](/home/gettingstarted/0.7/generators/index.html)).
+
+      public final JGen<Integer> sizes = JGen.intValues("size", 500000);
+    
+      public final JGen<Tuple2<Integer, LinkedList<Integer>>> lists = sizes.zip(sizes.map(
+          new Fun1<Integer, LinkedList<Integer>>() {
+            public LinkedList<Integer> apply(Integer v) {
+              return new LinkedList<Integer>();
+            }
           }
+      ));
+      
+      @gen("lists")
+      @benchmark("list")
+      public void histogram(Tuple2<Integer, LinkedList<Integer>> v) {
+        
+      }
 
-Finally, we declare the code snippet for the test by implementing the `snippet` method:
+Now we initialize our list with random data in the setup method:
 
-    public HashMap<Integer, LinkedList<Integer>> snippet(LinkedList<Integer> in) {
-      HashMap<Integer, LinkedList<Integer>> hm = new HashMap<Integer, LinkedList<Integer>>();
+      public void listSetup(Tuple2<Integer, LinkedList<Integer>> v) {
+        int size = v._1();
+        LinkedList<Integer> list = v._2();
+    
+        Random random = new Random(size);
+        for (int i = 0; i < size; i++) {
+          list.add(random.nextInt(size));
+        }
+      }
+      
+      @gen("lists")
+      @setup("listSetup")
+      @benchmark("list")
+      public void histogram(Tuple2<Integer, LinkedList<Integer>> v) {
+        
+      }
+      
+Next, we would like to have custom curve name:
```

Comment:
By default, the curve corresponding to the benchmark snippet will be equal to the method name.
    If we would like a custom name for this method, we need to use the `@curve` annotation: